### PR TITLE
WIP draft of JdbiTemplate based plugin

### PR DESCRIPTION
**Pull request by alwins0n**
This PR addresses #983 but also #987 

The idea is to follow springs conventions as closely as possible and expose a narrow interface that, similar to springs JdbcOperations, solely relies on the framework for transaction management.

If this draft is approved I will add tests and better documentation.

Also it can be considered to make `JdbiUtil` non-public.

Furthermore I was thinking of supporting spring boot which uses a lot more classpath magic to even avoid writing java config

### CODE REVIEW

**Comment by alwins0n** [comment]
jeez what happened? ran the tests fine locally... or so i thought. will check it out later

File: spring4/src/main/java/org/jdbi/v3/spring4/JdbiOperations.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+    /**
+     * like {@link Jdbi#withHandle(HandleCallback)} but respects springs transactions
+     */
+    <R, X extends Exception> R withHandle(HandleCallback<R, X> callback) throws X;
+
+    /**
+     * like {@link Jdbi#useHandle(HandleConsumer)} but respects springs transactions
+     */
+    default <X extends Exception> void useHandle(HandleConsumer<X> callback) throws X {
```
Comment (arteam) [inline spring4/src/main/java/org/jdbi/v3/spring4/JdbiOperations.java]: I think it would be better to implement `useHandle` instead of reusing `withHandle`. Having a callback in a callback will make stack traces in case of errors more confusing.

File: spring4/src/test/java/org/jdbi/v3/spring4/TestJdbiTemplate.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    @Autowired
+    private JdbiOperations jdbiOps;
+
+    @Test
+    public void testJdbiOps() {
+        assertThat(jdbiOps).isNotNull().isInstanceOf(JdbiTemplate.class);
+    }
+
+    @Configuration
+    @EnableTransactionManagement
```
Comment (arteam) [inline spring4/src/test/java/org/jdbi/v3/spring4/TestJdbiTemplate.java]: I believe to make `@EnableTransactionManagement` work you need to define a transaction manager in the configuration. In our case `DataSourceTransactionManager`:
```java
@Bean
public PlatformTransactionManager transactionManager(DataSource dataSource) {
    return new DataSourceTransactionManager(dataSource);
}
```

File: spring4/src/main/java/org/jdbi/v3/spring4/JdbiTemplate.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+    /**
+     * like {@link Jdbi#withHandle(HandleCallback)} but respects springs transactions
+     */
+    @Override
+    public <R, X extends Exception> R withHandle(HandleCallback<R, X> callback) throws X {
+        Handle h = JdbiUtil.getHandle(jdbi);
+        R result = callback.withHandle(h);
+        // close if not in transaction
+        JdbiUtil.closeIfNeeded(h);
```
Comment (arteam) [inline spring4/src/main/java/org/jdbi/v3/spring4/JdbiTemplate.java]: I believe we shouldn't close the handle here. It will be closed in the `afterCompletion` method. I think also it will probably be ignored because the handle is still binded to the active transaction.

File: spring4/src/test/java/org/jdbi/v3/spring4/TestJdbiTemplate.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        public Jdbi jdbi(DataSource dataSource) {
+            return Jdbi.create(dataSource);
+        }
+
+        @Bean
+        public JdbiOperations jdbiOperations(Jdbi jdbi) {
+            return new JdbiTemplate(jdbi);
+        }
+
+    }
```
Comment (arteam) [inline spring4/src/test/java/org/jdbi/v3/spring4/TestJdbiTemplate.java]: It would be nice to add a test how to use Dao class with Spring transactions and `JdbiTemplate`. Something like:
```java
        @Bean
        public Dao dao(JdbiOperations jdbiOperations) {
            return new Dao(jdbiOperations);
        }

    }

    public static class Dao {

        private JdbiOperations jdbiOperations;

        public Dao(JdbiOperations jdbiOperations) {
            this.jdbiOperations = jdbiOperations;
        }

        @Transactional
        public Integer getAmount() {
            return jdbiOperations.withHandle(h -> h.createQuery("select count(*) from something")
                    .mapTo(Integer.class)
                    .findOnly());
        }

        @Transactional
        public void insert(int id, String name) {
            jdbiOperations.useHandle(h -> h.createUpdate("insert into something (id, name) values (?, ?)")
                    .bind(0, id)
                    .bind(1, name)
                    .execute());
        }

        @Transactional
        public void insertMany(List<Integer> ids, List<String> names) {
            for (int i = 0; i < ids.size(); i++) {
                insert(ids.get(i), names.get(i));
            }
        }
    }

    @Test
    public void testTransactionsWork() {
        dao.insertMany(Arrays.asList(1, 2, 3), Arrays.asList("Test1", "Test2", "Test3"));
        System.out.println(dao.getAmount());
    }
```

File: spring4/src/main/java/org/jdbi/v3/spring4/JdbiOperations.java (hunk 1 above)
Comment (alwins0n) [reply spring4/src/main/java/org/jdbi/v3/spring4/JdbiOperations.java]: Didn't think about that, but yeah, can do

File: spring4/src/main/java/org/jdbi/v3/spring4/JdbiTemplate.java (hunk 3 above)
Comment (alwins0n) [reply spring4/src/main/java/org/jdbi/v3/spring4/JdbiTemplate.java]: that's point - just because you use JdbiTemplate does not guarantee that you are within a spring transaction (most likely still). 

closeIfNeeded closes the handle if it's NOT transaction bound, so exactly what I want here.

File: spring4/src/test/java/org/jdbi/v3/spring4/TestJdbiTemplate.java (hunk 2 above)
Comment (alwins0n) [reply spring4/src/test/java/org/jdbi/v3/spring4/TestJdbiTemplate.java]: maybe, will look that up. I am used to spring boot where the TM is defaulty defined as long as as a `PlatformTransactionManager` is on the classpath. 

then again the test suite is far from complete - this is WIP

File: spring4/src/test/java/org/jdbi/v3/spring4/TestJdbiTemplate.java (hunk 4 above)
Comment (alwins0n) [reply spring4/src/test/java/org/jdbi/v3/spring4/TestJdbiTemplate.java]: I will add checks to verify the behaviour I expect from the plugin. this is WIP

**Comment by alwins0n** [comment]
I should add the following: in this PR the Transaction Adapter for spring was modified to close the handle upon completion (as it should imho). I don't think this plugin was used a lot in the wild, but still - this change would maybe modify the behaviour of existing applications.

Either we should not directly expose JdbiUtil - ppl who upgrad the jdbi3-spring would have to deal with this (and maybe replace by JdbiTemplate) or leave public and deprecate the methods as they should not be used directly anymore (and the user will be notified of the change in behaviour)

File: spring4/src/main/java/org/jdbi/v3/spring4/JdbiOperations.java (hunk 1 above)
Comment (qualidafial) [reply spring4/src/main/java/org/jdbi/v3/spring4/JdbiOperations.java]: Actually.. `Handle.useHandle()` is already delegating to `withHandle()` :man_shrugging:

**Comment by alwins0n** [comment]
fyi: do not merge this already. found some cases where tests will fail. I am trying to understand spring and jdbi transaction management better. At this point I think a custom `PlatformTransactionManager` for Jdbi that does not operate on the DataSource directly but on Jdbi's Handles might be a clean and correct solution. Will update this PR asap

File: spring4/src/main/java/org/jdbi/v3/spring4/JdbiOperations.java (hunk 1 above)
Comment (alwins0n) [reply spring4/src/main/java/org/jdbi/v3/spring4/JdbiOperations.java]: true that, will leave it that way

File: spring4/src/main/java/org/jdbi/v3/spring4/JdbiTemplate.java (hunk 3 above)
Comment (alwins0n) [reply spring4/src/main/java/org/jdbi/v3/spring4/JdbiTemplate.java]: became obsolete with latest push

**Comment by alwins0n** [comment]
ough... after re-reading spring docs and digging into the transaction SPI as well as jdbi's interface for transactions - I decided to give a fully blown `JdbiTransactionManager` a go!

I pushed the latest version of this draft and its kind of a big one. I do not yet feel confident about everything but I think this is the way to go integrating spring with jdbi.

I tried to reuse as much convienience (Template classes etc.) from spring as possible.

Some points that need to be addressed:
- I rewrote lots of `JdbiUtil`, it breaks the old tests and implementations depending on it. After re-reading the code I think the present logic of JdbiUtil should be packed into `JdbiTemplate` and leave the old implementation be - and deprecate the api, if you feel that we should.
- Now that we have a jdbi `PlatformTransactionManager` we can invoke framework transactions also programmatically. I will show this in the future docs
- Docs need a full rewrite. I intend keeping part of the old docs because the setup with `JdbiFactoryBean` can be very nice/convenient in xml-config. If you approve these changes/PR I will update the docs accordingly

Merry Christmas btw

**Comment by qualidafial** [comment]
I would definitely prefer deprecating JdbiUtil over making an incompatible change.

**Comment by alwins0n** [comment]
Thanks for your comment. I will take that into account once this change is approved on principal. I would also invite ppl to question my approach. The reason I did this larger change (with respect to the existing implementation) was that I think if springs `DatasourceTransactionManager` operates directly on the Ds and its connections - Jdbi might not know what is going on. That way e.g. `Handle`s would be left unclosed after a transaction (where the underlying `Connection` will be closed) and might fail/behave unexpectedly when operated on them.

But maybe I am missing something and this cannot occur anyways and thus I am overcomplicating things here. Please provide feedback! Should we try to get someone from spring framework to comment here? About this being the "right" or "clean" approach?

File: spring4/src/main/java/org/jdbi/v3/spring4/JdbiTransactionObject.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+import org.springframework.transaction.TransactionSystemException;
+
+/**
+ * savepoint capable jdbi transaction wrapper
+ */
+public class JdbiTransactionObject implements SavepointManager {
+
+    private static final String JDBI_SAVEPOINT_BASENAME = "JDBI_SAVEPOINT_";
+
+    private int nextSavePointNumber = 0;
```
Comment (stevenschlansker) [inline spring4/src/main/java/org/jdbi/v3/spring4/JdbiTransactionObject.java]: should this be an `AtomicInteger`?

File: spring4/src/main/java/org/jdbi/v3/spring4/JdbiTransactionObject.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    private HandleHolder handleHolder;
+
+    public JdbiTransactionObject(HandleHolder handleHolder) {
+        this.handleHolder = handleHolder;
+    }
+
+    @Override
+    public Object createSavepoint() throws TransactionException {
+        String savePointName = JDBI_SAVEPOINT_BASENAME + nextSavePointNumber++;
```
Comment (stevenschlansker) [inline spring4/src/main/java/org/jdbi/v3/spring4/JdbiTransactionObject.java]: because this might be called from multiple concurrent threads, I'd guess

File: spring4/src/test/java/org/jdbi/v3/spring4/TestJdbiTemplate.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        @Bean
+        public DataSource dataSource() {
+            return new EmbeddedDatabaseBuilder()
+                    .setType(EmbeddedDatabaseType.H2)
+                    .build();
+        }
+
+        @Bean
+        public Jdbi jdbi(DataSource dataSource) {
+            return Jdbi.create(dataSource);
```
Comment (stevenschlansker) [inline spring4/src/test/java/org/jdbi/v3/spring4/TestJdbiTemplate.java]: What if this instead created a `DataSouce` implementation that does the `JdbiTemplate` / `JdbiUtil` logic, and delegates to the `dataSource` parameter?

File: spring4/src/main/java/org/jdbi/v3/spring4/JdbiUtil.java [hunk 8]
```diff
@@ -55,38 +53,9 @@ public static Handle getHandle(Jdbi jdbi)
      */
     public static void closeIfNeeded(Handle handle)
     {
-        if (!TRANSACTIONAL_HANDLES.contains(handle))
+        if (UNMANAGED_HANDLES.contains(handle))
```
Comment (stevenschlansker) [inline spring4/src/main/java/org/jdbi/v3/spring4/JdbiUtil.java]: this looks like a memory leak, should this be `remove`?

File: spring4/src/main/java/org/jdbi/v3/spring4/JdbiTemplate.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    private final Jdbi jdbi;
+
+    public JdbiTemplate(Jdbi jdbi) {
+        this.jdbi = jdbi;
+    }
+
+    /**
+     * like {@link Jdbi#withHandle(HandleCallback)} but respects springs transactions
+     */
+    @Override
```
Comment (stevenschlansker) [inline spring4/src/main/java/org/jdbi/v3/spring4/JdbiTemplate.java]: Ok, I think I am starting to see the intent here.  But I wonder, it looks like the only purpose of `JdbiTemplate` is to modify the behavior of `withHandle`.  Instead of introducing it as a wrapper with an interface, what if we wrap the `ConnectionFactory` that gets called in `Jdbi.open()` to invoke the `JdbiUtil.getHandle`?  And for the close logic, you could hook the `StatementBuilder.close()` logic

Then there is no need for this wrapper class, I think.  Does that sound like it might work?

File: spring4/src/main/java/org/jdbi/v3/spring4/JdbiTransactionObject.java (hunk 5 above)
Comment (alwins0n) [reply spring4/src/main/java/org/jdbi/v3/spring4/JdbiTransactionObject.java]: hmm. on principal the transaction and any transaction related objects are afaik thread bound (at least in the local context of a spring application). that means no more than one thread will ever increment the number of save points.

I will look that up though and make it safe for general purposes if necessary

File: spring4/src/main/java/org/jdbi/v3/spring4/JdbiTemplate.java (hunk 9 above)
Comment (alwins0n) [reply spring4/src/main/java/org/jdbi/v3/spring4/JdbiTemplate.java]: Interesting, let me think this through loudly:

If the connectionfactory is "spring transaction aware" the callstack of using jdbi within a spring transaction would approximately look like this

`Jdbi.useHandle` -> `JdbiUtil.getHandle` -> doStuffWith handle -> `Handle.close` -> `StatementBuilder.close` and `Connection.close`

everything after the "do Stuff" I find problematic. 
- The handle does not know anything about about the transaction
- `StatementBuilder.close` could do the right thing, thats okay!
- the `java.sql.Connection` is closed (Line 132) - bad!
- the handle is marked as closed (released) which may or may not be correct!
- ...

also:
- what if you used `Jdbi.open directly` - you now must not close it but use `JdbiUtil.closeIfNeeded`. are you supposed to know that?
- what if you create a transaction manually? `Jdbi.useTransaction` and so on
...

I have had these concerns before and thats why I decided 
- to wrap the calls to Jdbi via transaction aware classes - so Jdbi is properly informed about every change in the transactions and the handle state reflects the state in the transaction.
- to discourage the use of transaction methods on `Jdbi` - you can start a transaction programmatically via springs mechanisms
- to expose a spring-like API for convenience

If you see any possibility to go without the API  - I am happy to implement it that way. But I am not really happy with static calls and Jdbi not knowing about springs state and vice versa

File: spring4/src/test/java/org/jdbi/v3/spring4/TestJdbiTemplate.java (hunk 7 above)
Comment (alwins0n) [reply spring4/src/test/java/org/jdbi/v3/spring4/TestJdbiTemplate.java]: That's okay, but see my comment above. I think in this case spring manipulates stuff and state which Jdbi does not know anything about. also other problems I mentioned arise.

**Comment by alwins0n** [comment]
Can I get a ping if this is still of interest? 

The way I see it you are not happy with introducing another API but agree with the need of having a custom `PlatformTransactionManager` that keeps jdbi and spring in sync. Am I correct?

Regarding the API I investigated whether `JdbiTemplate` could be removed.

- It might be possible to modify the `Jdbi` class in a way that makes a direct usage possible, but it would be heavy modifications. Even if `Jdbi.open()` has a hook for getting the transaciton handle. `Jdbi.withHandle()` closes it after usage and would need a lot of hook code too, etc...
- If you use the plugin you are a Spring developer. I claim here, that Spring developers have no problem using a `xxxTemplate` class to achieve whatever they want in a "spring" way. e.g. have a look at: http://www.buggybread.com/2015/03/spring-framework-list-of-template.html . I would also be willing to remove the `JdbiOperations` interface altough its convention to provide one (for proxying, stubbing, testing)

Btw I noticed that the old version is also not in sync with the spring-style naming conventions for static access utility functions. `JdbiUtil` should be `JdbiUtils` analogue to `DatasourceUtils`, `LdapUtils`, ...

**Comment by stevenschlansker** [comment]
Yes, still of interest!  I'll do more reviewing shortly.

**Comment by alwins0n** [comment]
just wanted to say that I am actively using this PR in a real world project right now. Thus, I would really appreciate feedback. Right now it is running quite well.

I realize that this PR is quite big and the incentive to dig into spring is quite low. I can lay out my design decisions once more to have them challenged properly. Maybe some sequence diagrams might help to understand my reasoning.

Would love to hear how this is going forward.

**Comment by stevenschlansker** [comment]
Never fear, you aren't forgotten.  You're right that Spring support isn't top of our list but we definitely intend to move forward and testing the PR in a real world project is great progress.

**Comment by alwins0n** [comment]
hey guys, I want to move this forward - providing API methods for SQL object (useExtension, onDemand) on the high level approach (JdbiTemplate) as well as tests and updated docs.

Also spring 5 arrived which should be taken into consideration to support primarily. And autoconfigurations for spring boot 2 could be implemented.

should we try to get someone from spring on board to review this? should I publish it standalone? what's the move here?

**Comment by stevenschlansker** [comment]
OK, I don't see any reason to hold it up further - if you want to do another pass at the docs, and I noticed a couple of TODO test cases, then we can merge it!

**Comment by qualidafial** [comment]
@alwins0n I'm curious if `TransactionAwareDataSourceProxy` (whew!) mentioned on #1023 made any difference. You had mentioned you were going to try it out and see if it spares us the complexity of wrapping the transactions ourselves.

**Comment by alwins0n** [comment]
Haven't tried it out yet, but as I said, for the most use cases it will work properly. Again: it might be a problem if `Handle` and `Connection` are not in sync due to this "quickfix" solution. 

Off the top of my head, let me give you an example:
I request a handle from `Jdbi` somewhere within a spring transaction (e.g. annotated method). I do stuff with it and leave the annotated method. Now spring will close the underlying connection of this transaction. `Jdbi`s `Handle` does not know about this. If i plan to use it again (another call to the database) I will go fail (closed unexpected... etc)

On the other hand:
I request (open) a handle from `Jdbi` somewhere within a spring transaction (e.g. annotated method). Now i close it (try finally) since this is what I think I am expected to do. This works, but after control flow leaves the annotated method spring will attempt to close/release the connection -> this (might) fail and results in exceptions (probably).

I did not do thourough tests. Also example 1 might be a corner case (using a handle in one method and passing it out to reuse). But it seemed to me from the beginning that `TransactionAwareDataSourceProxy` and the likes are a hack! it might work most of the time but the proper way to integrate spring and jdbi is to teach spring that `Handle` is the transactional resource to be synced, not the `Connections`.

I know that the code above is scary to non-"springers" (hell, to everyone actually) but I think a library should provide the proper solution, not the hack, even if that means somewhat bloated code.

To wrap up - I will run some tests against the `TransactionAwareDataSourceProxy` to give myself a clear picture. but the mismatch of spring and jdbi as i mentioned will never go away. It just might be that it's not as "bad" as suspected by me

**Comment by alwins0n** [comment]
Can I ask for a decision here? IMHO either:
- document the usage of `TransactionAwareDataSourceProxy` in the plugin docs and make users aware that `Handle` and the underlying connection may not be in sync (e.g. closed). This might not pose a problem if JDBI is purly used in "lambda mode"
- update the plugin based on this PR which is a clean solution but requires the introduction of a Spring-like API ( `JdbiTemplate`)

**Comment by qualidafial** [comment]
Hi @alwins0n, I just realized I never got back to you.

I don't feel like we have enough information to make a call here. On your March 23 comment, you said you were planning to try out `TransactionAwareDataSourceProxy` and see where that gets us. I'm still waiting to hear your insights from that experiment.

Speaking personally, I am a little gun shy about adding Spring integration APIs, because I don't want to screw it up. It's hard for me to judge what's idiomatic in Spring because I don't use it--I use Dropwizard at my day job.

So I'm relying rather heavily on you and other community members who care to weigh in on what the best course of action should be.

If you've already posted your findings and I've lost track of it, please link me to it.

**Comment by alwins0n** [comment]
Well, I used my current test suite to run against a `TransactionAwareDataSourceProxy` wiring and it worked fine. That are my results as of now.

It is probably the best and simplest solution to document that configuration and leave the plugin be, I have to admit.

We should just warn that passing around handles to other methods (which may trigger leaving springs transaction context) or creating transactions on jdbi manually while being in springs transaciton context may lead to unexpected results.

as long as you rely on spring for transaction management and use jdbi's handle purely in "callback" mode (which is 99% of all cases) you should be fine.

Some minor improvements to the plugin could be made anyway as I mentioned on the whole thread here (naming etc.)

**Comment by alwins0n** [comment]
I am, but where? Is it really part of the plugin documentation? It feels more like some blog-post content (how to integrate jdbi with spring....)

**Comment by stevenschlansker** [comment]
I was imagining it'd have a whole section in the main jdbi docs, perhaps near the end with other "addons".
Also, if we are willing to tag the module as `@Beta` or `@Experimental`, that'd go a long way towards easing my worries about keeping a stable api.

File: spring4/src/main/java/org/jdbi/v3/spring4/JdbiUtil.java [hunk 10]
```diff
@@ -13,20 +13,19 @@
...
+import java.util.HashSet;
+import java.util.Set;
+
 /**
  * Utility for working with Jdbi and Spring transaction bound resources
  */
 public class JdbiUtil
 {
-    private static final Set<Handle> TRANSACTIONAL_HANDLES = new HashSet<>();
+    private static final Set<Handle> UNMANAGED_HANDLES = new HashSet<>();
```
Comment (osi) [inline spring4/src/main/java/org/jdbi/v3/spring4/JdbiUtil.java]: This set should be either synchronized or a set made from a concurrent hash map

**Comment by stevenschlansker** [comment]
Just checking in -- does anyone use this code in production / how's testing going?  Do you still intend to continue with the checklist of TODOs above?

**Comment by stevenschlansker** [comment]
Sounds like we're going to close this PR for now, but if anyone is interested in carrying the flag across the finish line we can always re-open it.
